sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReduceNumShufflePartitions.scala

```diff
@@ -91,14 +92,25 @@ case class ReduceNumShufflePartitions(conf: SQLConf) extends Rule[SparkPlan] {
       val distinctNumPreShufflePartitions =
         validMetrics.map(stats => stats.bytesByPartitionId.length).distinct
       if (validMetrics.nonEmpty && distinctNumPreShufflePartitions.length == 1) {
-        val partitionStartIndices = estimatePartitionStartIndices(validMetrics.toArray)
         // This transformation adds new nodes, so we must use `transformUp` here.
-        plan.transformUp {
+        plan.transformDown {
           // even for shuffle exchange whose input RDD has 0 partition, we should still update its
           // `partitionStartIndices`, so that all the leaf shuffles in a stage have the same
           // number of output partitions.
-          case stage: QueryStageExec if ShuffleQueryStageExec.isShuffleQueryStageExec(stage) =>
-            CoalescedShuffleReaderExec(stage, partitionStartIndices)
+          case stage: QueryStageExec if (ShuffleQueryStageExec.isShuffleQueryStageExec(stage) &&
+            !stage.visited) =>
+            stage.visited = true
```

Comment:
can we create a local set to keep the id of visited stages? Then we don't need to make `QueryStage` mutable.